numpyro/contrib/control_flow/cond.py

```diff
@@ -0,0 +1,131 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from functools import partial
+
+from jax import device_put, lax
+
+from numpyro import handlers
+from numpyro.contrib.control_flow.util import PytreeTrace
+from numpyro.primitives import _PYRO_STACK, apply_stack
+
+
+def _subs_wrapper(subs_map, site):
+    if isinstance(subs_map, dict) and site["name"] in subs_map:
+        return subs_map[site["name"]]
+    elif callable(subs_map):
+        rng_key = site["kwargs"].get("rng_key")
+        subs_map = (
+            handlers.seed(subs_map, rng_seed=rng_key)
+            if rng_key is not None
+            else subs_map
+        )
+        return subs_map(site)
+    return None
+
+
+def wrap_fn(fn, substitute_stack):
+    def wrapper(wrapped_operand):
+        rng_key, operand = wrapped_operand
+
+        with handlers.block():
+            seeded_fn = handlers.seed(fn, rng_key) if rng_key is not None else fn
+            for subs_type, subs_map in substitute_stack:
+                subs_fn = partial(_subs_wrapper, subs_map)
+                if subs_type == "condition":
+                    seeded_fn = handlers.condition(seeded_fn, condition_fn=subs_fn)
+                elif subs_type == "substitute":
+                    seeded_fn = handlers.substitute(seeded_fn, substitute_fn=subs_fn)
+
+            with handlers.trace() as trace:
+                value = seeded_fn(operand)
+
+        return value, PytreeTrace(trace)
+
+    return wrapper
+
+
+def cond_wrapper(
+    pred, true_fun, false_fun, operand, rng_key=None, substitute_stack=None
+):
+    if substitute_stack is None:
+        substitute_stack = []
+
+    wrapped_true_fun = wrap_fn(true_fun, substitute_stack)
+    wrapped_false_fun = wrap_fn(false_fun, substitute_stack)
+    wrapped_operand = device_put((rng_key, operand))
+    return lax.cond(pred, wrapped_true_fun, wrapped_false_fun, wrapped_operand)
+
+
+def cond(pred, true_fun, false_fun, operand):
+    """
+    This primitive conditionally applies `true_fun` or `false_fun`. See
+    :func:`jax.lax.cond` for more information.
```

Comment:
Could you add a `note` section to mention that all `sample` sites need to belong to the same distribution class?

Reply:
Good suggestion, have added something.